examples/experimental/omniglot_protonet_pytorch/model_def.py

```diff
@@ -0,0 +1,240 @@
+
+        def conv_block(in_channels, out_channels):
+            return nn.Sequential(
+                nn.Conv2d(in_channels, out_channels, 3, padding=1),
+                nn.BatchNorm2d(out_channels),
+                nn.ReLU(),
+                nn.MaxPool2d(2),
+            )
+
+        return nn.Sequential(
+            conv_block(x_dim, hid_dim),
+            conv_block(hid_dim, hid_dim),
+            conv_block(hid_dim, hid_dim),
+            conv_block(hid_dim, z_dim),
+            Flatten(),
+        )
+
+    def loss(self, x_support, y_support, x_query, y_query, model, split):
+        # x dimension N x C x H x W
+        _, channels, height, width = x_support.size()
+
+        num_classes = self.num_classes[split]
+        num_support = int(y_support.size(0) / num_classes)
+        num_query = int(y_query.size(0) / num_classes)
+
+        # First resort x so examples are ordered by class.
+        support_inds = torch.argsort(y_support)
+        query_inds = torch.argsort(y_query)
+        x_support = x_support[support_inds]
+        y_support = y_support[support_inds]
+        x_query = x_query[query_inds]
+        y_query = y_query[query_inds]
+
+        # Group support and query data into one forward pass
+
+        x = torch.cat([x_support, x_query], 0)
+
+        embedding = model(x)
+        embedding_dim = embedding.size(-1)
+
+        # Now we can reshape to get prototype embeddings
+        # Prototype size: (num_classes, embedding_dim)
+        prototypes = (
+            embedding[0 : num_classes * num_support]
+            .view(num_classes, num_support, embedding_dim)
+            .mean(1)
+        )
+
+        # Embedded query size: (num_classes * num_query, embedding_dim)
+        embedded_query = embedding[num_classes * num_support :]
+
+        # Compute distance between query set and prototypes
+        # Distance size: (num_classes * num_query, num_classes)
+        euclidean_dist = SquaredDistance(embedded_query, prototypes)
+
+        # Class log probabilities by treating -distances as logits
+        # Log_prob_query size: (num_classes, num_query, num_classes)
+        log_prob_query = F.log_softmax(-euclidean_dist, dim=1).view(
+            num_classes, num_query, -1
+        )
+
+        # Match query examples with classes
+        y_query_expand = y_query.view(num_classes, num_query, 1)
+        loss = -log_prob_query.gather(2, y_query_expand).squeeze().view(-1).mean()
+
+        _, pred_query = log_prob_query.max(2)
+
+        acc = torch.eq(pred_query, y_query_expand.squeeze()).float().mean()
+
+        return loss, acc
+
+    def optimizer(self, model: nn.Module):
+        optimizer = torch.optim.Adam(
+            model.parameters(),
+            lr=self.context.get_hparam("learning_rate"),
+            weight_decay=self.context.get_hparam("weight_decay"),
+        )
+        return optimizer
+
+    def create_lr_scheduler(self, optimizer: torch.optim.Optimizer):
```

Comment:
move optimizer and create_lr_scheduler above loss(). hard to look back and forth for referencing.